Quick filters

shamoon opened the pull request:
Hey @jonaswinkler I got inspired form our discussion in (closes #110) . This PR includes a pretty functional and speedy UI like we talked about.

- Quick access to Tags / Correspondents / Document Types. All lists scroll if they're very long and can be filtered using optional search field
- Can build up filter components using click to toggle on / off
- Quick "Clear" button
- Hooked up to existing components so using this UI correctly sets the "Filter Editor" (I named it that =) and then can still be saved as a view, etc.
- Video below (I fixed a couple small things after video was made).
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/4887959/101612281-530f9900-39bf-11eb-9e61-ecbd4da15090.gif)

Welcome your thoughts & testing!

jonaswinkler commented:
Did some head scratching.

- I'd like to replace the entire filter editor with that. Make it full width, add a text field for title filtering, and drop downs for date selection. That's pretty much all the filter editor is offering at this point anyway.

![image](https://user-images.githubusercontent.com/17569239/101623550-e9bc7400-3a18-11eb-88e4-b45bf9ca348b.png)

- I'd move all widgets into the filter-editor, and replace what's already there. I still like the idea of hiding the filtering controls on saved views.

- I'd move the entire filter rule editing logic into the filter editor (we can access child components with some `@ViewChild()` annotation, have to look up how that works)

- Remove apply button =minus one needed click.

- rxjs/debounce the entire thing, so that typing in the text field only filters the view when the user hasn't typed anything for a set amount of time.

- Add class="shadow" to the drop downs. It's nice.

- Allow multiple tags to be selected.

As for the date selection, two options.

Either something like this:

![image](https://user-images.githubusercontent.com/17569239/101623162-5125f400-3a18-11eb-94c0-b395768d1e41.png)

Or a date drill down in that menu, which first shows available years, then months, then days. Needs API support. Don't want to fetch all documents just to calculate that info.

-------

Do you think this is a good idea? I wasn't that happy with the filter editor anyway and especially found the date editing stuff borderline usable.

If you want to take a stab at it, notify me. I can take care of the date editing stuff afterwards. Or I'll just merge this and go at it myself.

jonaswinkler commented:
Edit.

Thinking further about the date drill down, I'd say it might be a good idea for all filters to only show available options (i.e., options that don't return an empty query set when clicked). Please don't do that in the front end though, I'd rather have some property in the API responses that tells us available options.

shamoon commented:
Hmm, the thing is the idea behind this PR was to have “quick” access to these things. I find the fact that they’re hidden inside the filter editor to be very cumbersome from a UI standpoint (I’d guess others do to) and it’s not ideal to require the user to have saved a view in order to do this. My suspicion is that with quick filters like these there will be will be at least some people who use them a majority of the time rather than anything else. Again, closer to the GitHub style. I agree those date filter changes are improvements just wasn’t really the intent of this PR.

Obviously you’re the boss it’s you’re repo :) and it’s tough without a lot of input from others but as a front-end person that’s my gut feeling.

shamoon commented:
Ps. I guess even just making the filter editor always visible would sort of be another solution?

shamoon commented:
Ps. I guess even just making the filter editor always visible would sort of be another solution (especially if we get rid of the apply button)?

I can take a crack at the other stuff once we decide this part

jonaswinkler commented:
Oh yeah, it seems I forgot to mention that. When the entire filter UI is this compact, I see no issue with always displaying that.

shamoon commented:
Making some good progress, screenshot attached. Gotta figure out the date ones, add debouncing on the dropdown filters, probably some other small things

<img width="1432" alt="Screen Shot 2020-12-10 at 3 48 35 PM" src="https://user-images.githubusercontent.com/4887959/101843528-3f6b4c00-3aff-11eb-8cdc-1c6bd72194fd.png">

jonaswinkler commented:
This is awesome!

The debounce thing is only really important for the text field. I think it would be nice if the UI would just update when you're done typing, instead of having to explicitly press enter or a button.

Regarding the date dropdowns, I think it would be really nice if you could select from a list of available years first, which would limit the view to that year. Then the contents of the dropdown would change to the list of months for which documents are available, and a button to deselect the year, and so on. But even if you just put the date range selectors in there [see 1], this is already a huge improvement.

Maybe `class="sticky-top"` the filter thing, don't know if that's useful.

==> I see some badges, I figure that's the number of documents with that tag? Do you download the entire document list to figure that out?

[1] https://ng-bootstrap.github.io/#/components/dropdown/examples#form

shamoon commented:
Glad you like it! Yea thats exactly how I implemented debounce on the search field. Yea the badges are just taken from `document_count`. In general I agree this would all be better if the options were updated "live" (like only tags with count > 0 are shown as you build up the filter) but I think this works for now and its built such that when the API is ready it shouldnt be too painful to integrate.

shamoon commented:
So this code is in pretty good shape and things work well, please take a look!

jonaswinkler commented:
Will do!

jonaswinkler commented:
Nice work! I'll have a look at the code later.

For now, I've spotted an issue with multiple tags:
- select a tag A 
- select another tag B
- deselect B
- The view does not update and B is still in the filter rules.

jonaswinkler commented on src-ui/src/app/services/filter-editor-view.service.ts:
```diff
@@ -0,0 +1,188 @@
+    return this.documentTypes?.filter(dt => documentTypeRules.find(dtr => dtr.value == dt.id))
+  }
+
+  toggleFitlerByTag(tag: PaperlessTag) {
```
Typo.

jonaswinkler replied:
More below.

shamoon replied:
Sorry where's the typo here?

jonaswinkler replied:
Fitler, Filter

shamoon replied:
Oh haha I didn't even notice! Fixed 😄

jonaswinkler commented on src-ui/src/app/services/filter-editor-view.service.ts:
```diff
@@ -0,0 +1,188 @@
+@Injectable({
+  providedIn: 'root'
+})
+export class FilterEditorViewService {
```
I like this. We can set filter rules from anywhere and not be concerned about how it's done.

jonaswinkler commented on src-ui/src/app/services/filter-editor-view.service.ts:
```diff
@@ -0,0 +1,188 @@
+export class FilterEditorViewService {
+  tags$: Observable<Results<PaperlessTag>>
+  correspondents$: Observable<Results<PaperlessCorrespondent>>
+  documentTypes$: Observable<Results<PaperlessDocumentType>>
```
Any reason we keep both the observables and the lists?

shamoon replied:
Great point, refactored to just pass the lists to components

jonaswinkler commented on src-ui/src/app/pipes/filter.pipe.ts:
```diff
@@ -0,0 +1,17 @@
+@Pipe({
+  name: 'filter'
+})
+export class FilterPipe implements PipeTransform {
```
Ingenious and versatile. I like it.

jonaswinkler commented on src-ui/src/app/components/filter-editor/filter-editor.component.html:
```diff
@@ -1,52 +1,22 @@
-      <option [ngValue]="true">Yes</option>
-      <option [ngValue]="false">No</option>
-    </select>
+  <app-filter-dropdown class="col-auto" [(items$)]="filterEditorService.tags$" [itemsSelected]="filterEditorService.selectedTags" [title]="'Tags'" [display]="'tag'" (toggle)="onToggleTag($event)"></app-filter-dropdown>
```
`title="Tags"`, and below.

shamoon replied:
Actually I think needs to be `'Tags'` to pass a string literal =) https://stackoverflow.com/questions/47057696/how-to-use-es6-template-literal-as-angular-component-input

jonaswinkler reviewed (commented):
Apart from a couple things here and there, this looks alright to me.

Please address that issue with the tags and I'll have it merged!

shamoon commented:
Ok sorted that bug about toggling items. That logic was tricky 😁

shamoon commented:
Actually I want to make a couple more small improvements, will let you know when they're sorted soon!

jonaswinkler commented:
If you're already at it, here's the things I'd do after merging:
- Make the filter fields in the drop downs always visible on top, only make the items scrollable
- Probably remove the class="card" border, I don't think that's needed, have to try and see how it looks and feels
- Provide a clear / remove / show all button in the date drop downs, since its not so easy to clear the date fields
- At least for the "this week", "this year" etc, use the same menu item style as in all the other drop down menus
- Experiment with closing the drop downs after selection. I can see the use of leaving them open for the tags drop down, but for all the others, a user will most likely inspect the list after selecting a filter option, and will almost always want to close the drop down.

shamoon commented:
Ok! This is looking fancy, I like it. I added all that you mentioned with the exception of the auto-closing. To me it feels unexpected for the menus to auto-close, even with correspondents where you can only toggle one. I think the user will look at / scroll the list and pick one and may want to pick a different one more often than wanting to close the menu. You can still of course click outside to close it. The one exception for which I added auto-closing was when they use the filtering to narrow the list down to a single item and then press Enter. This toggles the item and closes the menu. Feels pretty good to me but let me know your thoughts!

shamoon commented:
Screenshot 😀
<img width="1367" alt="Screen Shot 2020-12-13 at 3 49 39 PM" src="https://user-images.githubusercontent.com/4887959/102027703-d2ec8900-3d5a-11eb-9f4b-5f2d7fd7e7a5.png">

jonaswinkler commented:
Thanks again for all the effort.

I don't see anything major and will merge this, if you're happy with the results. I'll probably make some adjustments to the code here and there.

shamoon commented:
Haha just thought of one more thing I’d like to attempt a little later! Thanks for your patience

jonaswinkler commented:
You know, there will always be a couple things left to do ;)

shamoon commented:
Ha, yes always... Well the thing I was going to add was to be able to search Title & Content at the same time. I built that into the service (not committed) but then I realized the filters are additive so it searches for "title has ___" AND "content has ___". Seems like would need a backend change to make it work, like maybe if the search string is the same for title & content do an OR? Perhaps theres a better way Im missing.

Anyway, otherwise I think it is actually good to go!

jonaswinkler commented:
That workaround for the infinite loop is really nasty.

shamoon commented:
> That workaround for the infinite loop is really nasty.

I know... 😵

jonaswinkler commented:
Just to wrap this up, here's what I've changed.

- In the end, I've decided that the filter view service is not required. Filter state should only permanently managed by one instance in the application, and that is the document list service. This also manages all other aspects of the document list.
- The filter editor should just be given a filter rule set, modify that, and don't remember anything across page navigation, hence no need for a backing service. This makes a couple things fairly easier.
- It was actually causing issues when navigating between saved views and the dashboard.
- The badges indicating active filters are nice, but the changes in width of the buttons not so much. I'm switching between `btn-outline-primary` and `btn-primary`, since this is what all the other buttons do as well to indicate that something is active.

shamoon commented:
Thanks for the explanation, all makes sense. Yea I debated about the badges. What do you think about the attached, putting the badge in the top right, it doesnt cause reflow but its kinda cool to see how many are selected IMHO.
<img width="344" alt="Screen Shot 2020-12-14 at 7 52 13 PM" src="https://user-images.githubusercontent.com/4887959/102168935-ffc99a80-3e45-11eb-9fa3-adbfcb0b6f79.png">
